File: src/rpc_demo.erl
-module(rpc_demo).
-export([
         test/0,
         json_call/2,
         bert_call/5,
         pb_call/2, pb_call/3
        ]).
-include_lib("eunit/include/eunit.hrl").
-include("rpc_demo_pb.hrl").

test() ->
    application:start(inets),
    inets:start(httpc, [{profile, default}]),
    eunit:test(rpc_demo).

webmachine_test() ->
    Expected = sequence:sequence(5),
    Actual = json_call("http://localhost:8001/sequence/5"),
    ?assertEqual(Expected, Actual).

spooky_test() ->
    Expected = sequence:sequence(5),
    Actual = json_call("http://localhost:8002/sequence/5"),
    ?assertEqual(Expected, Actual).

bert_test() ->
    Expected = sequence:sequence(5),
    Actual = bert_call("localhost", 9999, ernie_sequence, sequence, [5]),
    ?assertEqual(Expected, Actual).

pb_test() ->
    Expected = sequence:sequence(5),
    {ok, Response} = pb_call("localhost", 8003, #sequencerequest { n = 5 }),
    Actual = {ok, Response#sequenceresponse.sequence},
    ?assertEqual(Expected, Actual).

%%% Helper Functions %%%

json_call(Url) ->
    json_call(Url, default).
json_call(Url, Profile) ->
    Headers = [
               {"Content-Type", "application/json"},
               {"Connection", "Keep-Alive"}
              ],
    Request = {Url, Headers},
    case httpc:request(get, Request, [], [], Profile) of
        {ok, {{_, 200, _}, _, Body}} ->
            {ok, mochijson2:decode(Body)};
        Other ->
            {error, Other}
    end.

bert_call(Host, Port, Mod, Fun, Args) ->
    case gen_tcp:connect(Host, Port, [binary, {packet, 4}]) of
        {ok, Sock} ->
            bert_call_1(Sock, Mod, Fun, Args);
        Other ->
            io:format("Unable to establish connection: ~p~n", [Other])
    end.
bert_call_1(Sock, Mod, Fun, Args) ->
    RequestBytes = bert:encode({call, Mod, Fun, Args}),
    ok = gen_tcp:send(Sock, RequestBytes),
    receive
        {tcp, _Port, ResponseBytes} ->
            ok = gen_tcp:close(Sock),
            case bert:decode(ResponseBytes) of
                {reply, Reply} ->
                    Reply;
                Other ->
                    {error, {unexpected_response, Other}}
            end;
        {tcp_closed, Port} ->
            io:format("Connection closed after sending data: ~p~n", [Port]),
            {error, {tcp_closed, Port}};
        Other ->
            ok = gen_tcp:close(Sock),
            io:format("Unexpected message: ~p~n", [Other]),
            {error, Other}
    end.


pb_call(Host, Port, Req) ->
    case gen_tcp:connect(Host, Port, [binary, {packet, 4}, {header, 1}]) of
        {ok, Sock} ->
            Result = pb_call(Sock, Req),
            gen_tcp:close(Sock),
            Result;
        Other ->
            io:format("Unable to establish connection: ~p~n", [Other])
    end.
pb_call(Sock, Req) ->
    RequestBytes = protobuff_server:encode(Req),
    ok = gen_tcp:send(Sock, RequestBytes),
    receive
        {tcp, _Port, ResponseBytes} ->
            {ok, protobuff_server:decode(ResponseBytes)};
        {tcp_closed, Port} ->
            io:format("Connection closed after sending data: ~p~n", [Port]),
            {error, {tcp_closed, Port}};
        Other ->
            io:format("Unexpected message: ~p~n", [Other]),
            {error, Other}
    end.

File: lib/rpc_demo.rb
#!/usr/bin/env ruby
require 'rubygems'
require 'net/http'
require 'json'
require 'bertrpc'
require 'beefcake'
require 'socket'

def webmachine_test()
  url = 'http://localhost:8001/sequence/5'
  resp = Net::HTTP.get_response(URI.parse(url))
  sequence = JSON.parse(resp.body)
  puts "Webmachine: #{sequence}"
end

def spooky_test()
  url = 'http://localhost:8002/sequence/5'
  resp = Net::HTTP.get_response(URI.parse(url))
  sequence = JSON.parse(resp.body)
  puts "Spooky: #{sequence}"
end

def bert_test()
  svc = BERTRPC::Service.new('localhost', 9999)
  sequence = svc.call.ernie_sequence.sequence(5)[1]
  puts "BERT: #{sequence}"
end

class SequenceRequest
  include Beefcake::Message
  MsgCode = 1
  required :n, :int32, 1
end

class SequenceResponse
  include Beefcake::Message
  MsgCode = 2
  repeated :sequence, :string, 1
end

class SequenceError
  include Beefcake::Message
  MsgCode = 3
  repeated :message, :string, 1
end

def pb_test()
  Socket.tcp("localhost", 8003) do |socket|
    req = SequenceRequest.new(:n => 5)
    write_protobuff(socket, SequenceRequest::MsgCode, req)
    resp = read_protobuf(socket)
    puts "PB: #{resp.sequence}"
  end
end

def write_protobuff(socket, message_code, message)
  encoded = message.encode
  header = [encoded.length + 1, message_code].pack("NC")
  socket.write(header + encoded)
end

def read_protobuf(socket)
  header = socket.read(5)
  size, code = header.unpack("NC")
  message = socket.read(size-1)
  case code
  when SequenceResponse::MsgCode
    SequenceResponse.decode(message)
  when SequenceError::MsgCode
    SequenceError.decode(message)
  end
end

webmachine_test()
spooky_test()
bert_test()
pb_test()
